File: Kotlin/linked-list-in-binary-tree.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun isSubPath(head: ListNode?, root: TreeNode?): Boolean {
        if (head == null) {
            return true
        }
        if (root == null) {
            return false
        }
        return comparePath(head, root) || isSubPath(head, root?.left) || isSubPath(head, root?.right)
    }

    fun comparePath(head: ListNode?, root: TreeNode?): Boolean {
        if (head == null) {
            return true
        }
        if (root == null) {
            return false
        }
        return head.`val` == root.`val` && (comparePath(head?.next, root?.left) || comparePath(head?.next, root?.right))
    }
}

File: Go/linked-list-in-binary-tree.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSubPath(head *ListNode, root *TreeNode) bool {
	if head == nil {
		return true
	}
	if root == nil {
		return false
	}
	return comparePath(head, root) || isSubPath(head, root.Left) || isSubPath(head, root.Right)
}

func comparePath(head *ListNode, root *TreeNode) bool {
	if head == nil {
		return true
	}
	if root == nil {
		return false
	}
	return head.Val == root.Val && (comparePath(head.Next, root.Left) || comparePath(head.Next, root.Right))
}